### PR TITLE
*: LOAD DATA support WITH DETACHED

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref #40499

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

```
MySQL [test]> load data infile 's3://load-csv/test.customer.aaaa.csv?access_key=xxx&secret_access_key=xxx' ignore into table my_table FIELDS TERMINATED BY ',' OPTIONALLY ENCLOSED BY '' LINES TERMINATED BY '\\n' IGNORE 1 LINES with detached;
+--------+
| Job_ID |
+--------+
|  90001 |
+--------+
1 row in set (2.651 sec)

MySQL [test]> select * from mysql.load_data_jobs where job_id = 90001\G
*************************** 1. row ***************************
         job_id: 90001
expected_status: running
    create_time: 2023-03-13 17:15:17.887803
     start_time: 2023-03-13 17:15:17.890134
    update_time: 2023-03-13 17:15:25.945554
       end_time: 2023-03-13 17:15:25.958597
    data_source: s3://load-csv/test.customer.aaaa.csv?access_key=xxx&secret_access_key=xxx
   table_schema: test
     table_name: my_table
    import_mode: logical
    create_user: root@%
       progress: {"SourceFileSize":52428118,"LoadedFileSize":52428118,"LoadedRowCnt":89281}
 result_message: Records: 89281  Deleted: 0  Skipped: 0  Warnings: 65535
  error_message: NULL
```

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: executor/load_data.go [hunk 1]
```diff
@@ -245,6 +270,7 @@ type LoadDataWorker struct {
 	progress        atomic.Pointer[asyncloaddata.Progress]
 	getSysSessionFn func() (sessionctx.Context, error)
 	putSysSessionFn func(context.Context, sessionctx.Context)
+	ownSession      bool
```
Comment (D3Hunter) [inline executor/load_data.go:273]: reuse `detached`? always has same value with it

File: executor/load_data.go [hunk 2]
```diff
@@ -358,14 +392,48 @@ func (e *LoadDataWorker) Load(
...
+}
+
+func (e *LoadDataWorker) doLoad(
+	ctx context.Context,
+	readerInfos []LoadDataReaderInfo,
+	jobID int64,
+) (err error) {
+	defer func() {
+		if e.ownSession {
+			e.putSysSessionFn(ctx, e.Ctx)
```
Comment (D3Hunter) [inline executor/load_data.go:419]: maybe remove `e.Ctx`, we're using `e.ctx` and `e.Ctx` mixingly, it's same as ctx in `InsertValues`

File: executor/load_data.go [hunk 3]
```diff
@@ -358,14 +392,48 @@ func (e *LoadDataWorker) Load(
...
+		}
+	}()
+
+	s, err := e.getSysSessionFn()
```
Comment (D3Hunter) [inline executor/load_data.go:423]: why get a new session to execute job related sql?

File: executor/load_data.go [hunk 4]
```diff
@@ -358,14 +392,48 @@ func (e *LoadDataWorker) Load(
...
 	if err != nil {
 		return err
 	}
+	defer e.putSysSessionFn(ctx, s)
+
+	sqlExec := s.(sqlexec.SQLExecutor)
 
 	defer func() {
+		ctx2 := context.Background()
```
Comment (D3Hunter) [inline executor/load_data.go:432]: make it work even ctx is cancelled? maybe leave a comment

File: executor/load_data.go (hunk 2 above)
Comment (lance6716) [reply executor/load_data.go:419]: it's made public for LOAD DATA LOCAL where the clientConn.handleLoadData check Ctx.GetSessionVars().Killed to handle reading packet. Maybe I can unify the CANCEL LOAD DATA JOB with this checking killed in next PR.

File: executor/load_data.go (hunk 4 above)
Comment (lance6716) [reply executor/load_data.go:462]: https://github.com/pingcap/tidb/pull/42159/commits/4fa316edede8af4b70dd34949066f6967563a808

**Review by D3Hunter — APPROVED** [review]

File: planner/core/planbuilder.go [hunk 5]
```diff
@@ -4238,6 +4248,12 @@ func (b *PlanBuilder) buildLoadData(ctx context.Context, ld *ast.LoadDataStmt) (
 	if err != nil {
 		return nil, err
 	}
+
+	if detached {
```
Comment (D3Hunter) [inline planner/core/planbuilder.go:4252]: can we move this into `buildLoadData`, so no need to init `detached

Comment (lance6716) [reply planner/core/planbuilder.go:4252]: This is already buildLoadData. You mean buildLoadData (executor)?

executor builder can not change the schema of plan, it's one way

Comment (lance6716) [reply planner/core/planbuilder.go:4252]: I moved it into buildLoadData (executor) and the query schema is not changed. Maybe other components have already read the old value and decided the query output.

File: executor/load_data.go [hunk 6]
```diff
@@ -249,12 +278,51 @@ type LoadDataWorker struct {
...
+		if err != nil {
+			return nil, err
+		}
+		// copy the related variables to the new session
+		// I have no confident that all needed variables are copied :(
+		from := userSctx.GetSessionVars()
+		to := sysSession.GetSessionVars()
+		to.User = from.User
+		to.CurrentDB = from.CurrentDB
+		to.SQLMode = from.SQLMode
```
Comment (Defined2014) [inline executor/load_data.go:322]: Maybe you could ref https://github.com/pingcap/tidb/blob/fff7fb9fecb1149df5982c302197227901ed1693/ddl/backfilling.go#L805 to determine which variable are useful.

Not sure timezone is useful for `load_data` or not.

Comment (lance6716) [reply executor/load_data.go:322]: I added a simple test for related variables https://github.com/pingcap/tidb/pull/42159/commits/0b82a0aa2fe5ef7cd54a81830501461b366c5a43

**Review by hawkingrei — APPROVED** [review]

**Comment by lance6716** [comment]
https://github.com/pingcap/tidb/pull/42159

**Pull request merged**